libraries/persistent_store/src/format.rs

```diff
@@ -0,0 +1,907 @@
+    /// entry, except in the checksum itself. So it needs 6 bits to store numbers between 0 and 58.
+    // NOTE: It may be possible to save one bit by storing:
+    // - the footer checksum (as a field) if the value is not empty
+    // - the header checksum (as a checksum) if the value is empty
+    HEADER_CHECKSUM: Checksum <= 58,
+
+    #[cfg(test)]
+    LEN_HEADER: Length,
+}
+
+// The fields of an erase entry.
+bitfield! {
+    /// The identifier for erase entries.
+    ID_ERASE: ConstField = [1 1 0 0 0],
+
+    /// The page to be erased.
+    ERASE_PAGE: Field <= MAX_PAGE,
+
+    #[cfg(test)]
+    LEN_ERASE: Length,
+}
+
+// The fields of a clear entry.
+bitfield! {
+    /// The identifier for clear entries.
+    ID_CLEAR: ConstField = [1 1 0 0 1],
+
+    /// The minimum key to be cleared.
+    ///
+    /// All entries with a key below this limit are not cleared. All other entries are deleted.
+    CLEAR_MIN_KEY: Field <= MAX_KEY,
+
+    #[cfg(test)]
+    LEN_CLEAR: Length,
+}
+
+// The fields of a marker entry.
+bitfield! {
+    /// The identifier for marker entries.
+    ID_MARKER: ConstField = [1 1 0 1 0],
+
+    /// The number of updates in this transaction.
+    ///
+    /// The update entries follow this marker entry.
+    MARKER_COUNT: Field <= MAX_UPDATES,
+
+    #[cfg(test)]
+    LEN_MARKER: Length,
+}
+
+// The fields of a remove entry.
+bitfield! {
+    /// The identifier for remove entries.
+    ID_REMOVE: ConstField = [1 1 0 1 1],
+
+    /// The key of the user entry to be removed.
+    REMOVE_KEY: Field <= MAX_KEY,
+
+    #[cfg(test)]
+    LEN_REMOVE: Length,
+}
+
+/// The position of a word in the virtual storage.
+///
+/// With the notations defined in `Format`, let:
+/// - `w` a virtual word offset in a page which is between `0` and `Q - 1`
+/// - `p` a page offset which is between `0` and `N - 1`
+/// - `c` the number of erase cycles of a page which is between `0` and `E`
+///
+/// Then the position of a word is `(c*N + p)*Q + w`. This position monotonically increases and
+/// represents the consumed lifetime of the storage.
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub struct Position(usize);
+
+impl core::ops::Add<usize> for Position {
+    type Output = Position;
+
+    fn add(self, delta: usize) -> Position {
+        Position(self.0 + delta)
+    }
+}
+
+impl core::ops::Sub<Position> for Position {
+    type Output = usize;
+
+    fn sub(self, base: Position) -> usize {
+        self.0 - base.0
+    }
+}
+
+impl core::ops::AddAssign<usize> for Position {
+    fn add_assign(&mut self, delta: usize) {
+        self.0 += delta;
+    }
+}
+
+impl Position {
+    /// Create a word position given its coordinates.
+    ///
+    /// The coordinates of a word are:
+    /// - Its word index in its page.
+    /// - Its page index in the storage.
+    /// - The number of times that page was erased.
+    pub fn new(format: &Format, cycle: usize, page: usize, word: usize) -> Position {
```

Comment:
The representation as `usize` is not for optimization purposes, it's for semantics reasons. The position of a word in the virtual storage is a natural number. Those functions permit to convert back and forth between a position in the linear virtual storage and a position in the physical storage (erase cycle is needed for the conversion to be bijective but it doesn't matter for the physical storage).

Regarding the more precise comments:
- Those 3 components form a lexicographical order and they are always given in order: cycle, page, word. We could prevent the risk of typo by introducing a struct just for the argument of `new`, but I don't think it's useful because getting the order wrong would have very obvious bad consequences.
- The non-trivial arithmetic is just how one converts from the cartesian product of finite ordered sets to an ordered set of same cardinal preserving order when the cartesian product is equipped with lexicographical order.
- If not, then `get` and all arithmetic on positions would be non-trivial (see the implementations above).

For `StorageIndex` it doesn't make sense to have a linear space, because operations do not span pages. Note however that for the in-memory implementation (used for tests and fuzzing) we still store it as a linear space, but for a real implementation this is not the case (in particular when the storage is split on multiple banks).

`Position` is mostly used in RAM. Only the difference between 2 positions is stored in the storage during compaction.

Reply:
> The position of a word in the virtual storage is a natural number.

Using a natural number is a possible representation, but semantically a position in virtual storage is IMO more meaningfully a (cycle, page, word) triplet. Position "123" doesn't make more sense than (cycle 3, page 2, word 42).

> Those 3 components form a lexicographical order and they are always given in order: cycle, page, word.

Sounds fair.

> The non-trivial arithmetic is just how one converts from the cartesian product of finite ordered sets to an ordered set of same cardinal preserving order when the cartesian product is equipped with lexicographical order.

I'm not denying that, just saying that it's unnecessary code complexity.

> If not, then `get` and all arithmetic on positions would be non-trivial (see the implementations above).

With 3 fields:
- `index`, `page_begin`, `next_page`, `cycle`, `page`, `word` become trivial,
- `get` doesn't get less trivial than the current `new` constructor, i.e. computing `(cycle * format.num_pages() + page) * format.virt_page_size() + word`,
- I'm not sure about the arithmetic. What's an acceptable range for the `delta`? Is any arithmetic useful, or do we just perform a subset of operations, e.g. increasing the cycle count or switching to a different page? Being able to compute `position + 123` doesn't seem like a robust API to me. For example, say `(0, 0, 0) + 123` would be `(0, 0, 123)` (same cycle, same page, different word) but `(0, 63, 127) + 123` would be `(1, 0, 122)` (different cycle, different page, different word). Or to put it differently, the delta has to be some combination of a cycle, page and word, right? In which case I think it's clearer to represent both `Position` and `Delta` as triplets of (cycle, page, word). But again, hard to tell at this point because this pull request doesn't show any use of the `Position` methods yet.

Reply:
> > The position of a word in the virtual storage is a natural number
>
> Using a natural number is a possible representation

It's not about representation. It's about semantics. The position of a word in the virtual storage is a natural number (I'm not writing "is represented by" but "is", which is a mathematical definition). The virtual storage is a linear space. Indexing in a linear space is done by natural numbers. The physical storage is a list of pages. Indexing a word in a list of pages is done with a tuple of natural numbers.

If it's hard to tell because pull requests are split in parts, you can always refer to the complete version of the code (including fuzzing and integration in OpenSK) [here](https://github.com/ia0/OpenSK/tree/v2/libraries/persistent_store). But those concepts should hopefully be already clarified in the top-level documentation in `lib.rs`.

Reply:
I replied twice that this is about semantics and you still talk about representation. I'm not sure if it's because you don't understand the design or because you really think the representation should not follow the design.

Regarding design, either you did not read the [first step](https://github.com/google/OpenSK/pull/172/files#diff-02b7313abaae70c974fb90b50b6469eac3b249bf94b4b751595acef6771b63a0R143-R151) of the design, or you don't understand it. Please clarify.

Regarding representation, I think you are still missing the fact that we do arithmetic on positions although I said it in my first reply. For example when we [increment](https://github.com/ia0/OpenSK/blob/v2/libraries/persistent_store/src/store.rs#L875) the position to get the position of the next entry. This is done many times in particular in `parse_entry` which is a core function. We also do [subtraction](https://github.com/ia0/OpenSK/blob/v2/libraries/persistent_store/src/store.rs#L211) to compute distance between positions.

If after that this is still unclear to you, you can simply try to change the representation and see what you would need to change to make the code compile and pass fuzzing. In particular, you will realize that for those arithmetic operations on positions you will need the format as argument which hints that you are breaking abstraction layers, because operations that are purely at the virtual storage layer should not have to understand the physical storage description.

Reply:
> I replied twice that this is about semantics and you still talk about representation. I'm not sure if it's because you don't understand the design or because you really think the representation should not follow the design.

Representation is an implementation choice that doesn't have to align with the design abstraction yes.

> Regarding design, either you did not read the [first step](https://github.com/google/OpenSK/pull/172/files#diff-02b7313abaae70c974fb90b50b6469eac3b249bf94b4b751595acef6771b63a0R143-R151) of the design, or you don't understand it. Please clarify.

I'll check again more thoroughly, but I understand the gist of it, in particular the section you pointed at.

Some things that strike me are that:
- positions that are more than `N * Q` (+ `Q`-aligned) words apart (i.e. belong to different erase cycles) cannot exist simultaneously,
- positions are at most `(E + 1) * N * Q` (due to the at most `E` erase cycles).

Therefore, operations on positions cannot really be decoupled from the format, which defines E, N and Q.

> Regarding representation, I think you are still missing the fact that we do arithmetic on positions although I said it in my first reply. For example when we [increment](https://github.com/ia0/OpenSK/blob/v2/libraries/persistent_store/src/store.rs#L875) the position to get the position of the next entry. This is done many times in particular in `parse_entry` which is a core function. We also do [subtraction](https://github.com/ia0/OpenSK/blob/v2/libraries/persistent_store/src/store.rs#L211) to compute distance between positions.

Of course, yet these could maybe be simpler to implement than the current modulo/division operations.

> If after that this is still unclear to you, you can simply try to change the representation and see what you would need to change to make the code compile and pass fuzzing. In particular, you will realize that for those arithmetic operations on positions you will need the format as argument which hints that you are breaking abstraction layers, because operations that are purely at the virtual storage layer should not have to understand the physical storage description.

Sure, addition and subtraction would need to be aware of the format configuration. But either all of the `index`/`page_begin`/`next_page`/`cycle`/`page`/`word` functions need to know the store, or `get`/addition/subtraction need to know the format. In the end, all of the relevant operations happen in functions of the `Store`, which already knows the `Format` so I'm not convinced by your argument that this breaks abstraction layers.

---

All in all, the implementation details of `Position` are not that big an issue given that they are behind the `Position` abstraction, but I was just pointing out that the current representation might be overly complex for what we do with it. Again, it's also been clearer after you pointed out the actual use cases as implemented in https://github.com/ia0/OpenSK/tree/v2/libraries/persistent_store.

Reply:
> Representation is an implementation choice that doesn't have to align with the design abstraction yes.

This is what I call optimization. And for this first version I would prefer to avoid optimizations until we know we need them.

> Therefore, operations on positions cannot really be decoupled from the format

Those are not operations but invariant. The first leave in the realm of code and syntax. The second leave in the realm of maths and semantics.

> All in all, the implementation details of Position are not that big an issue

Ok, resolving this conversation then.